### PR TITLE
refactor(expr): introduce MarkedArrayRef and eval_v2

### PR DESCRIPTION
**Pull request by TennyZhuang**
Signed-off-by: TennyZhuang <zty0826@gmail.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)
- Add the 'user-facing changes' label if your PR contains changes that are visible to users (optional)

## Checklist

- [ ] I have written necessary docs and comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/common/src/array/iterator.rs [hunk 1]
```diff
@@ -82,6 +82,35 @@ impl<'a> Iterator for ArrayImplIterator<'a> {
...
+    vis: &'a Vis,
+    arr: &'a ArrayImpl,
+    pos: usize,
+}
+
+impl<'a> Iterator for MarkedArayRefIter<'a> {
+    type Item = DatumRef<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while self.pos < self.arr.len() {
```
Comment (skyzh) [inline src/common/src/array/iterator.rs:95]: This would be hard for compiler to optimize :(

Comment (TennyZhuang) [reply src/common/src/array/iterator.rs:95]: I guess that we need array-bench :)

File: src/expr/src/expr/expr_input_ref.rs [hunk 2]
```diff
@@ -54,6 +54,12 @@ impl Expression for InputRefExpression {
         }
     }
 
+    fn eval_v2<'a>(&self, input: &'a DataChunk) -> Result<MarkedArrayRef<'a>> {
+        let arr = input.column_at(self.idx).array_ref();
+        let vis = input.vis();
+        Ok(MarkedArrayRef::from_ref(arr, vis))
+    }
```
Comment (fuyufjh) [inline src/expr/src/expr/expr_input_ref.rs:61]: It seems that the only usage of `MarkedArrayRef` is to preserve the `vis` array in `input: &DataChunk`, so why not just let `eval()` be **unaware** of `vis`? 

```rust
fn eval_v2<'a>(&self, input: &[ArrayImpl]) -> Result<ArrayImpl> {
```

Comment (TennyZhuang) [reply src/expr/src/expr/expr_input_ref.rs:61]: It will be wasteful, aware of vis is necessary to skip some heavy computation.

Comment (fuyufjh) [reply src/expr/src/expr/expr_input_ref.rs:61]: Yes, my mistake. Then how about

```
fn eval_v2<'a>(&self, input: &[ArrayImpl], vis: &Vis) -> Result<ArrayImpl> {
```

Comment (TennyZhuang) [reply src/expr/src/expr/expr_input_ref.rs:61]: It’s almost same as current eval.

Of course I know it work, but I think the interface is unsafe. We should not guarantee what will the output value be if the input vis = false. I want to investigate some safe interface.

**Comment by TennyZhuang** [comment]
Closed, `MarkedArrayRef` will cause that visibility to be handled twice in `a + b`, and even more in more complicated expressions. An unsafe `eval` method seems unavoidable here.

**Pull request closed without merging**